rpm/src/main/java/org/eclipse/packager/rpm/parse/InputHeader.java

```diff
@@ -58,45 +61,111 @@ public long getLength() {
+    }
+
+    @Override
+    public Long getLong(T tag) {
+        if (!Long.class.isAssignableFrom(tag.getDataType())) {
+            throw new IllegalArgumentException("Tag " + tag  + " is not a long");
+        }
+
+        return getOptionalTag(tag, Long.class).flatMap(headerValue -> headerValue.getValue().asLong()).orElse(null);
+    }
+
+    @Override
+    public List<String> getStringList(T tag) {
+        if (!String[].class.isAssignableFrom(tag.getDataType())) {
+            throw new IllegalArgumentException("Tag " + tag  + " is not an array of strings");
+        }
+
+        return getOptionalTag(tag, String[].class).flatMap(headerValue -> headerValue.getValue().asStringArray().map(Arrays::asList)).orElse(null);
     }
 
     @Override
-    public Optional<Object> getValue(final T tag) {
-        return Optional.ofNullable(getTag(tag));
+    public List<Integer> getIntegerList(T tag) {
+        if (!Integer[].class.isAssignableFrom(tag.getDataType())) {
+            throw new IllegalArgumentException("Tag " + tag  + " is not an array of integers");
+        }
+
+        return getOptionalTag(tag, Integer[].class).flatMap(headerValue -> headerValue.getValue().asIntegerArray().map(Arrays::asList)).orElse(null);
     }
 
-    public Optional<Object> getOptionalTag(final T tag) {
-        return getEntry(tag).map(HeaderValue::getValue);
+    @Override
+    public List<Long> getLongList(T tag) {
+        if (!Long[].class.isAssignableFrom(tag.getDataType())) {
+            throw new IllegalArgumentException("Tag " + tag  + " is not an array of longs");
+        }
+
+        return getOptionalTag(tag, Long[].class).flatMap(headerValue -> headerValue.getValue().asLongArray().map(Arrays::asList)).orElse(null);
     }
 
-    public Optional<Object> getOptionalTag(final int tag) {
-        return getEntry(tag).map(HeaderValue::getValue);
+    @Override
+    public byte[] getByteArray(T tag) {
+        if (!byte[].class.isAssignableFrom(tag.getDataType())) {
+            throw new IllegalArgumentException("Tag " + tag  + " is not an array of bytes");
+        }
+
+        return getOptionalTag(tag, byte[].class).flatMap(headerValue -> headerValue.getValue().asByteArray()).orElse(null);
     }
 
-    public Optional<HeaderValue> getEntry(final T tag) {
-        return Optional.ofNullable(this.entries.get(tag.getValue()));
+    public <E> Optional<HeaderValue<E>> getOptionalTag(final int tag, Class<E> dataType) {
+        return getEntry(tag, dataType);
     }
 
-    public Optional<HeaderValue> getEntry(final int tag) {
-        return Optional.ofNullable(this.entries.get(tag));
+    public <E> Optional<HeaderValue<E>> getOptionalTag(final T tag, Class<E> dataType) {
+        return getOptionalTag(tag.getValue(), dataType);
     }
 
-    public Object getTagOrDefault(final T tag, final Object defaultValue) {
-        return getOptionalTag(tag).orElse(defaultValue);
+    @SuppressWarnings("unchecked")
```

Comment:
Same for the next method, I think we should add an actual type check. IIRC, you can use `dataType.cast(..)` for this. However IIRC using `(HeaderValue<E>`) won't work, because that's simply being erased and becomes `HeaderValue`.

Reply:
OK, I will try to use `.cast()` here.

Reply:
Actually, now we're just returning `HeaderValue` with no type, so there's nothing to cast, right?

Reply:
Using:

```java
<E> Optional<HeaderValue<E>> getOptionalTag(final T tag, Class<E> dataType);
```

would actually have a benefit if the type of the value really is being limited to `E`. That requires casting using `dataType.cast(..)`. It would also be required to handle the case when the value is not of type `E`. Throw an exception, return `None` would both work.

The question is just if that makes the code calling into it much better?

Given Java's new pattern type construct (not sure I got the name right) one should be able to do something like:

```java
if  (headers.getEntry(Tags.MyTag) instanceof String s) {
 // do something with s
}
```

When reading, the header can basically carry anything, and we can't control that. Only when creating a header, we could enforce (via the type system) that we user cannot create wrong headers.